July-Morning/assignment1/Assignment_1_Q1.cpp

```diff
@@ -0,0 +1,91 @@
+#include <iostream>
+#include <string>
+#include <locale> 
+#include <unordered_map>
+#include <algorithm>
+
+// Check if two strings are anagrams (case sensitive algorithm)
+// Return value:
+// true - if strings are anagrams 
+// false - if strings are not anagrams or at least one string is empty 
+bool Is_Anagram (std::string s1, std::string s2)
+{
+    if ((s1.size() == 0) || (s2.size() == 0))
+        return false;
+    if (s1.size() != s2.size())
+        return false;
+    std::unordered_map<char, int> symbol_counter;
+    for (auto& sym : s1)
+        symbol_counter[sym]++;    
+    for (auto& sym : s2)
+        symbol_counter[sym]--;    
+    for (auto& sym : s2)

```

Comment:
That's a nice trick: going only over one of the strings characters is enough to check anagrams.
Another option is to go over the entries on the symbol counter:
```C++
for (const pair<char, int>& entry : symbol_counter) {
    if (entry.second != 0)
        return false;
}
```
This way, if the strings are very long (e.g. `str1 == str2 == aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa`), this loop goes only once per different character.
Also you don't need to look up every character in the map again (map lookup is O(log n) for std::map, and requires hash function to run in std::unordered_map. Not very expensive but still more than nothing.)
This does not improve the 'worst case' complexity but still may save some CPU cycles :)

P.S.
This loop is a good candidate to use `auto`:
```C++
for (const auto& entry : symbol_counter) {
    if (entry.second != 0)
        return false;
}
```